cmd/promtool/rules.go

```diff
@@ -0,0 +1,217 @@
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/rules"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+const maxSamplesInMemory = 5000
+
+type queryRangeAPI interface {
+	QueryRange(ctx context.Context, query string, r v1.Range) (model.Value, v1.Warnings, error)
+}
+
+type ruleImporter struct {
+	logger log.Logger
+	config ruleImporterConfig
+
+	apiClient queryRangeAPI
+
+	groups      map[string]*rules.Group
+	ruleManager *rules.Manager
+}
+
+type ruleImporterConfig struct {
+	outputDir    string
+	start        time.Time
+	end          time.Time
+	evalInterval time.Duration
+}
+
+// newRuleImporter creates a new rule importer that can be used to parse and evaluate recording rule files and create new series
+// written to disk in blocks.
+func newRuleImporter(logger log.Logger, config ruleImporterConfig, apiClient queryRangeAPI) *ruleImporter {
+
+	return &ruleImporter{
+		logger:      logger,
+		config:      config,
+		apiClient:   apiClient,
+		ruleManager: rules.NewManager(&rules.ManagerOptions{}),
+	}
+}
+
+// loadGroups parses groups from a list of recording rule files.
+func (importer *ruleImporter) loadGroups(ctx context.Context, filenames []string) (errs []error) {
+	groups, errs := importer.ruleManager.LoadGroups(importer.config.evalInterval, labels.Labels{}, filenames...)
+	if errs != nil {
+		return errs
+	}
+	importer.groups = groups
+	return nil
+}
+
+// importAll evaluates all the recording rules and creates new time series and writes them to disk in blocks.
+func (importer *ruleImporter) importAll(ctx context.Context) (errs []error) {
+	for name, group := range importer.groups {
+		level.Info(importer.logger).Log("backfiller", fmt.Sprintf("processing group, name: %s", name))
+
+		stimeWithAlignment := group.EvalTimestamp(importer.config.start.UnixNano())
+		for i, r := range group.Rules() {
+			level.Info(importer.logger).Log("backfiller", fmt.Sprintf("processing rule %d, name: %s", i+1, r.Name()))
+			if err := importer.importRule(ctx, r.Query().String(), r.Name(), r.Labels(), stimeWithAlignment, importer.config.end, group.Interval()); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	return errs
+}
+
+// importRule queries a prometheus API to evaluate rules at times in the past.
+func (importer *ruleImporter) importRule(ctx context.Context, ruleExpr, ruleName string, ruleLabels labels.Labels, start, end time.Time, interval time.Duration) (err error) {
+	blockDuration := tsdb.DefaultBlockDuration
+	startOfBlock := blockDuration * (start.Unix() / blockDuration)
+	for t := startOfBlock; t <= end.Unix(); t = t + blockDuration {
+		endOfBlock := t + blockDuration
```

Comment:
We probably want to calculate startOfBlock / start, and instead of executing a rule from the start of a block, delay it so we even even space results.